go/common/viewingkey/viewing_key.go

```diff
@@ -194,51 +224,53 @@ func CalculateUserID(publicKeyBytes []byte) []byte {
 }
 
 func VerifySignatureEIP712(userID string, address *gethcommon.Address, signature []byte, chainID int64) (bool, error) {
-	// get raw data for structured message
-	rawData, err := GenerateAuthenticationEIP712RawData(userID, chainID)
+	var rawDataOptions [][]byte
+
+	rawDataOptions, err := GenerateAuthenticationEIP712RawDataOptions(userID, chainID)
 	if err != nil {
 		return false, err
 	}
 
-	// create a hash of structured message (needed for signature verification)
-	hashBytes := crypto.Keccak256(rawData)
-	hash := gethcommon.BytesToHash(hashBytes)
+	for _, rawData := range rawDataOptions {
+		// create a hash of structured message (needed for signature verification)
+		hashBytes := crypto.Keccak256(rawData)
+		hash := gethcommon.BytesToHash(hashBytes)
 
-	if len(signature) != 65 {
-		return false, fmt.Errorf("invalid signature length: %d", len(signature))
-	}
+		if len(signature) != 65 {
+			continue
+		}
 
-	// We transform the V from 27/28 to 0/1. This same change is made in Geth internals, for legacy reasons to be able
-	// to recover the address: https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L452-L459
-	if signature[64] == 27 || signature[64] == 28 {
-		signature[64] -= 27
-	}
+		// We transform the V from 27/28 to 0/1. This same change is made in Geth internals, for legacy reasons to be able
+		// to recover the address: https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L452-L459
+		if signature[64] == 27 || signature[64] == 28 {
```

Comment:
There's quite a lot going on in this for loop, maybe we could reduce the nested complexity by extracting the attempted validation into a function.